src/libraries/Microsoft.PowerFx.Core/Public/CheckResult.cs

```diff
@@ -17,115 +19,413 @@
+
+        #region Results from IR 
+        private IRResult _irresult;
+
+        #endregion 
+
+        internal TexlBinding Binding
+        {
+            get
+            {
+                if (_binding == null)
+                {
+                    throw new InvalidOperationException($"Must call {nameof(ApplyBindingInternal)} before accessing binding.");
+                }
+
+                return _binding;
+            }
+        }
 
         /// <summary>
-        /// Results from parsing. 
+        /// List of all errors and warnings. Check <see cref="ExpressionError.IsWarning"/>.
+        /// This can include Parse, Bind, and per-engine custom errors (see <see cref="Engine.PostCheck(CheckResult)"/>, 
+        /// or any custom errors passes explicit to the ctor.
+        /// Not null, but empty on success.
         /// </summary>
-        public ParseResult Parse { get; set; }
+        public IEnumerable<ExpressionError> Errors
+        {
+            get => _errors;
+
+            [Obsolete("use constructor to set errors")]
+            set => _errors.AddRange(value);
+        }
 
         /// <summary>
-        /// Symbols passed to this binding. May be null. 
+        /// True if no errors for stages run so far. 
         /// </summary>
-        public ReadOnlySymbolTable Symbols { get; set; }
+        public bool IsSuccess => !_errors.Any(x => !x.IsWarning);              
+
+        /// <summary>
+        /// Helper to throw if <see cref="IsSuccess"/> is false.
+        /// </summary>
+        public void ThrowOnErrors()
+        {
+            if (!IsSuccess)
+            {
+                var msg = string.Join("\r\n", Errors.Select(err => err.ToString()).ToArray());
+                throw new InvalidOperationException($"Errors: " + msg);
+            }
+        }
+
+        internal bool HasDeferredArgsWarning => _errors.Any(x => x.IsWarning && x.MessageKey.Equals(TexlStrings.WarnDeferredType.Key));
+
+        /// <summary>
+        /// Full set of Symbols passed to this binding. 
+        /// Can include symbols from Config, Engine, and Parameters, 
+        /// May be null. 
+        /// </summary>
+        internal ReadOnlySymbolTable AllSymbols { get; private set; }
 
         /// <summary>
         /// Parameters are the subset of symbols that must be passed in Eval() for each evaluation. 
         /// This lets us associated the type in Check()  with the values in Eval().
         /// </summary>
-        internal ReadOnlySymbolTable Parameters { get; set; }
+        internal ReadOnlySymbolTable Parameters => _symbols;
 
         /// <summary>
         /// Culture info passed to this binding. May be null. 
         /// </summary>
-        internal CultureInfo CultureInfo { get; set; }
+        internal CultureInfo CultureInfo => this.Source.Config.CultureInfo;
+
+        internal void ThrowIfSymbolsChanged()
+        {
+            if (_symbols != null)
+            {
+                var endHash = _symbols.VersionHash;

```

Comment:
>endHash [](http://example.com/codeflow?start=20&length=7)

useless variable

Reply:
we can merge the two 'if' conditions for clarity

Reply:
This is a style thing. 
I prefer having the explicit locals because it's easier to spot things under a debugger and I can set breakpoints in the middle.